src/connectors/impls/gpubsub/consumer.rs

```diff
@@ -0,0 +1,404 @@
+use tonic::codegen::InterceptedService;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::Status;
+use tremor_common::blue_green_dashmap::BlueGreenDashMap;
+use tremor_pipeline::ConfigImpl;
+
+#[derive(Deserialize, Clone)]
+struct Config {
+    #[serde(default = "default_connect_timeout")]
+    pub connect_timeout: u64,
+    #[serde(default = "default_ack_deadline")]
+    pub ack_deadline: u64,
+    pub subscription_id: String,
+    #[serde(default = "default_endpoint")]
+    pub endpoint: String,
+    #[cfg(test)]
+    #[serde(default = "default_skip_authentication")]
+    pub skip_authentication: bool,
+}
+impl ConfigImpl for Config {}
+
+fn default_endpoint() -> String {
+    "https://pubsub.googleapis.com".into()
+}
+
+#[cfg(test)]
+fn default_skip_authentication() -> bool {
+    false
+}
+
+fn default_connect_timeout() -> u64 {
+    1_000_000_000u64 // 1 second
+}
+
+fn default_ack_deadline() -> u64 {
+    10_000_000_000u64 // 10 seconds
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "gpubsub_consumer".into()
+    }
+
+    async fn build(&self, alias: &str, raw_config: &ConnectorConfig) -> Result<Box<dyn Connector>> {
+        let client_id = format!("tremor-{}-{}-{:?}", hostname(), alias, task::current().id());
+
+        if let Some(raw) = &raw_config.config {
+            let config = Config::new(raw)?;
+            let url = Url::<HttpsDefaults>::parse(config.endpoint.as_str())?;
+
+            Ok(Box::new(GSub {
+                config,
+                url,
+                client_id,
+            }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(alias.to_string()).into())
+        }
+    }
+}
+
+struct GSub {
+    config: Config,
+    url: Url<HttpsDefaults>,
+    client_id: String,
+}
+
+type PubSubClient = SubscriberClient<InterceptedService<Channel, AuthInterceptor>>;
+type AsyncTaskMessage = Result<(u64, PubsubMessage)>;
+
+struct GSubSource {
+    config: Config,
+    client: Option<PubSubClient>,
+    receiver: Option<Receiver<AsyncTaskMessage>>,
+    ack_sender: Option<Sender<u64>>,
+    task_handle: Option<JoinHandle<()>>,
+    url: Url<HttpsDefaults>,
+    client_id: String,
+}
+
+impl GSubSource {
+    pub fn new(config: Config, url: Url<HttpsDefaults>, client_id: String) -> Self {
+        GSubSource {
+            config,
+            url,
+            client_id,
+            client: None,
+            receiver: None,
+            task_handle: None,
+            ack_sender: None,
+        }
+    }
+}
+
+async fn consumer_task(
+    mut client: PubSubClient,
+    client_id: String,
+    sender: Sender<AsyncTaskMessage>,
+    subscription_id: String,
+    ack_deadline: Duration,
+    ack_receiver: Receiver<u64>,
+) {
+    let mut ack_counter = 0;
+
+    let ack_ids = Arc::new(RwLock::new(BlueGreenDashMap::new(
```

Comment:
We also don't need the `BlueGreenDashMap` to be based on `DashMap` and can instead base it on `HashMap`, this might lower the synchronization overhead and go even more brrrrr.

Reply:
we do, because the async_stream generates it's own task

Reply:
But one of those synchronization layers could go away, right? As we need to mess with those two maps, i guess it is the RwLock, right? So we could base it on two `halfbrown::HashMap`s, couldn't we?

Reply:
good point, I switched to vanilla HashMap

Reply:
Not for today but it'd be intersting to benchmark the difference, I suspect the DashMap will be faster then the RwLock but it doubtfully matters in this context, just curiosity :)